pkg/probe/http/request.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/component-base/version"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/probe"
+)
+
+// NewProbeRequest returns an http.Request suitable for use as a request for a
+// probe.
+func NewProbeRequest(url *url.URL, headers http.Header) (*http.Request, error) {
+	return newProbeRequest(url, headers, "probe")
+}
+
+// NewRequestForHTTPGetAction returns an http.Request derived from httpGet.
+// When httpGet.Host is empty, podIP will be used instead.
+func NewRequestForHTTPGetAction(httpGet *v1.HTTPGetAction, container *v1.Container, podIP string, userAgentFrament string) (*http.Request, error) {
+	scheme := strings.ToLower(string(httpGet.Scheme))
+	if scheme == "" {
+		scheme = "http"
+	}
+
+	host := httpGet.Host
+	if host == "" {
+		host = podIP
+	}
+
+	var portStr string
+
+	hasPort := httpGet.Port.String() != ""
+	if hasPort || !utilfeature.DefaultFeatureGate.Enabled(features.ConsistentHTTPGetHandlers) {
```

Comment:
I thought that at first, too, but IIIRC, tests revealed conditions where this block was necessary.

It may be belt and braces, but `NewRequestForHTTPGetAction` is an exported function and it should not be assumed that there will be only one caller with well understood behaviors; although it's only called for lifecycle handlers now, there's no guarantee that will continue to remain the case. By checking the port here, users are protected from a deficiency of tests in case `NewREqeustForHTTPGetAction` is called in some other context, too.

> API validation requires the port field be populated for those anyway

Is it safe to assume that this code will only be called after validation? That seems like an unsafe assumption for an exported identifier.

Reply:
> IIIRC, tests revealed conditions where this block was necessary.

Hmm, I'd be curious to know if those were actual existing conditions… I really thought all paths that led here populated and validated port 

> By checking the port here, users are protected from a deficiency of tests in case NewREqeustForHTTPGetAction is called in some other context, too.

If we currently error if port is unset, I don't think we should relax that

Reply:
> Hmm, I'd be curious to know if those were actual existing conditions… I really thought all paths that led here populated and validated port

I'm not sure they were actual existing conditions, and I wouldn't be surprised if all code paths that lead here populate and validate the port. The conditions where it was necessary was just in blackbox testing `NewRequestForHTTPGetAction`.

> If we currently error if port is unset, I don't think we should relax that

Lifecycle handlers and probers were inconsistent. @tallclair identified the inconsistency in the [matrix of conditions](https://docs.google.com/spreadsheets/d/1YdmemfOj3jD7sVB5rme1QmbxB4vPI5YsHTdWY017l2o/edit#gid=0) that he graciously worked up. The change was further supported in the [refactor that he started](https://github.com/tallclair/kubernetes/commit/8fd8a3d823198e92a49347ae495900aa0cea09aa) to demonstrate his ideas. As far as I can tell, this wouldn't relax any validation that's done by Kubernetes, it only uses defaults in the exported functions that API clients could call. i.e. it appropriately uses an `HTTPGetAction` value and leaves validation to a higher layer.

I went ahead and added a fixup commit that addresses your concern here. I don't care either way, but please confirm whether you want to keep that commit or drop it after considering what the proper course of action is with the inconsistency between probes and lifecycle handlers that previously existed in mind.

Reply:
> As far as I can tell, this wouldn't relax any validation that's done by Kubernetes, it only uses defaults in the exported functions that API clients could call

These exported functions are in k8s.io/kubernetes, so we do not expect use of them as a library... the only caller we expect is the kubelet, whose inputs have already passed API validation.

I commented at https://github.com/kubernetes/kubernetes/pull/86139/files#r922627685 as well, but making lifecycle consistent with probers and requiring the port seems better.